### PR TITLE
[Refactor] Refactor multi-agent MLP

### PR DESCRIPTION
**Pull request by vmoens**
cc @matteobettini 

By the way, rather than doing a multi-agent MLP, a multi-agent ConvNet, a multi agent LSTM etc, should we not consider one way of batching ops in whichever nn.Module for multiagent and (maybe) just subclass that for every class?

Seems a bit repetitive no? Isn't there pare of the logic we can recycle?

TODO:
 - [ ] Use vmap and module ensemble for multiple nets

### CODE REVIEW

File: torchrl/modules/models/multiagent.py [hunk 1]
```diff
@@ -158,22 +162,38 @@ def __init__(
...
+            self.params = TensorDictParams(TensorDict.from_module(net), no_convert=True)
+        make_functional(net)
+        self.net = net
+        if self.centralised:
+            self.net_call = torch.vmap(
+                self.net,
+                (None, 0)
+            ) if not self.share_params else self.net
+        else:
+            self.net_call = torch.vmap(self.net, (-2, 0)) if not self.share_params else self.net
```
Comment (matteobettini) [inline torchrl/modules/models/multiagent.py:196]: Does this work and pass the test?

Cause it seems to me that we are not doing the resahping into (`n_agents` * `n_inputs_per_agent`) in the centralized case

Comment (vmoens) [reply torchrl/modules/models/multiagent.py:196]: No it doesn't. In fact it's not ready for review :)

Comment (vmoens) [reply torchrl/modules/models/multiagent.py:196]: But feel free to comment on the general idea from the PR description!

**Comment by matteobettini** [comment]
Because rnns in torchrl are TensorDictModules no?
In fact we do not have them in the `models`.

Yeah i think a common multiagent nn module can makes sense, we need to be parametric to the multiagent dim but it should be possible

File: torchrl/modules/models/multiagent.py [hunk 2]
```diff
@@ -158,22 +162,38 @@ def __init__(
...
+                **kwargs,
+            )
+
+        if not self.share_params:
+            agent_networks = [
+                make_net()
+                for _ in range(self.n_agents)
             ]
-        )
+            self.params = TensorDictParams(torch.stack([TensorDict.from_module(mod) for mod in agent_networks], 0).contiguous(), no_convert=True)
```
Comment (matteobettini) [inline torchrl/modules/models/multiagent.py:183]:
```suggestion
            self.params = TensorDictParams(torch.stack([TensorDict.from_module(make_net()) for _ in range(self.n_agents)], 0).contiguous(), no_convert=True)
```

**Comment by vmoens** [comment]
> Because rnns in torchrl are TensorDictModules no?
> In fact we do not have them in the models.

That can be easily worked around, it's up to us to adapt our design.

File: torchrl/modules/models/multiagent.py [hunk 3]
```diff
@@ -158,22 +163,45 @@ def __init__(
...
+        if self.centralised:
+            if not self.share_params:
+                self.net_call = torch.vmap(self.net, in_dims=(None, 0), out_dims=(-2,))
+            else:
+                self.net_call = self.net
+        else:
+            if not self.share_params:
+                self.net_call = torch.vmap(self.net, in_dims=(-2, 0), out_dims=(-2,))
+            else:
+                self.net_call = torch.vmap(self.net, in_dims=(-2, None), out_dims=(-2,))
```
Comment (matteobettini) [inline torchrl/modules/models/multiagent.py:202]: is this neded? wouldn't the mlp do vectorization by default?

Comment (matteobettini) [inline torchrl/modules/models/multiagent.py:202]:
```suggestion
        if self.share_params:
            self.net_call = self.net
        else:
            if self.centralised:
                self.net_call = torch.vmap(self.net, in_dims=(None, 0), out_dims=(-2,))
           else:
                self.net_call = torch.vmap(self.net, in_dims=(-2, 0), out_dims=(-2,))
```

Comment (vmoens) [reply torchrl/modules/models/multiagent.py:202]: Not sure all ops will support an arbitrary num of dims

Comment (matteobettini) [reply torchrl/modules/models/multiagent.py]: oh ok let's revert to previous version then and we are good to go

**Comment by vmoens** [comment]
The mappo test fails
I'm investigating that

**Comment by vmoens** [comment]
I think this is a good use case for first class dimension, as the bug is a bit convoluted to solve. Let me work on this and put this on hold

**Comment by kfu02** [comment]
Hi, I am just wondering if this issue is still being investigated, as I would like to try RNNs in the multi-agent setting (specifically for [MAPPO](https://arxiv.org/pdf/2103.01955.pdf), as that paper claims better performance for RNNs vs MLPs in many settings). Apologies if this is not the right place to ask.

**Comment by vmoens** [comment]
This is definitely a good place to ask. I can refresh and merge this PR now, and make a CNN/RNN version too. Bear with me...

**Comment by kfu02** [comment]
Great! I eagerly await the refresh/merge of this PR!

**Comment by vmoens** [comment]
@kfu02 

Check out https://github.com/pytorch/rl/pull/1921
It should work faster and allow us to code RNNs in a successive PR.
The issue right now is just that we can't account for non initialized lazy params as they are gathered in a single TensorDict during construction.

The solution will be a bit convoluted but doable :)
